// @mintlify/validation/dist/index.js
import { zodToJsonSchema } from 'zod-to-json-schema';
import { mintConfigSchema } from './mint-config/schemas/v1/config.js';
import { docsConfigSchema } from './mint-config/schemas/v2/index.js';
export * from './openapi/types/endpoint.js';
export * from './openapi/OpenApiToEndpointConverter.js';
export { stripComponents } from './openapi/stripComponents.js';
export { SchemaConverter } from './openapi/SchemaConverter.js';
export { generateExampleFromSchema } from './openapi/generateExampleFromSchema.js';
export { generateFirstIncrementalSchema, generateNextIncrementalSchema, } from './openapi/IncrementalEvaluator.js';
export { validateMintConfig } from './mint-config/validateMintConfig.js';
export { formatIssue } from './mint-config/formatIssue.js';
export { upgradeToDocsConfig } from './mint-config/upgrades/upgradeToDocsConfig.js';
export const mintConfigJsonSchema = zodToJsonSchema(mintConfigSchema, 'Schema');
export const docsConfigJsonSchema = zodToJsonSchema(docsConfigSchema, 'Schema');
export * from './mint-config/schemas/v2/properties/index.js';
