// @mintlify/scraping/bin/utils/log.js
export const colors = {
    red: '\x1b[31m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    magenta: '\x1b[35m',
    cyan: '\x1b[36m',
    default: '\x1b[0m',
};
const noColors = {
    red: '',
    green: '',
    yellow: '',
    blue: '',
    magenta: '',
    cyan: '',
    default: '',
};
const statuses = ['success', 'failure', 'error', 'warn', 'warning', 'info'];
export const activeColors = process.stdout.isTTY ? colors : noColors;
export const checkIcon = '✔ ';
export const xIcon = '✘ ';
export const infoIcon = 'ⓘ ';
export const warningIcon = '⚠ ';
export function log(message, statusOrColor) {
    let color = activeColors.blue;
    let statusMsg = 'INFO   ';
    let icon = infoIcon;
    const msg = typeof message === 'string' ? message.toLowerCase() : '';
    if (!statusOrColor) {
        statusOrColor =
            msg.includes('fail') ||
                msg.includes('error') ||
                msg.includes('could not') ||
                msg.includes("couldn't") ||
                msg.includes('did not') ||
                msg.includes('invalid') ||
                msg.includes("didn't")
                ? 'error'
                : msg.includes('success') ||
                    msg.includes('downloaded') ||
                    msg.includes('written') ||
                    msg.includes('added')
                    ? 'success'
                    : msg.includes('warn')
                        ? 'warn'
                        : undefined;
    }
    switch (statusOrColor) {
        case undefined:
        case 'info':
            break;
        case 'success':
            color = activeColors.green;
            statusMsg = 'SUCCESS';
            icon = checkIcon;
            break;
        case 'warn':
        case 'warning':
            color = activeColors.yellow;
            statusMsg = 'WARNING';
            icon = warningIcon;
            break;
        case 'failure':
        case 'error':
            color = activeColors.red;
            statusMsg = 'ERROR  ';
            icon = xIcon;
            break;
    }
    console.log(`${color}${icon} ${statusMsg}${activeColors.default} - ${typeof message === 'string' ||
        typeof message === 'bigint' ||
        typeof message === 'number' ||
        typeof message === 'boolean'
        ? message
        : JSON.stringify(message, undefined, 2)}`);
}
