// @mintlify/validation/dist/mint-config/schemas/v2/properties/navigation/tabs.js
import { z } from 'zod';
import { hiddenSchema } from '../reusable/hidden.js';
import { hrefSchema } from '../reusable/href.js';
import { iconSchema } from '../reusable/icon.js';
import { openApiSchema } from '../reusable/openapi.js';
import { anchorsSchema } from './anchors.js';
import { dropdownsSchema } from './dropdown.js';
import { groupsSchema } from './groups.js';
import { languagesSchema } from './languages.js';
import { pagesSchema } from './pages.js';
import { versionsSchema } from './version.js';
const baseTabSchema = z.object({
    tab: z.string().nonempty().describe('The name of the tab'),
    icon: iconSchema.optional(),
    hidden: hiddenSchema.optional(),
});
export const tabSchema = baseTabSchema.and(z.union([
    z.object({ openapi: openApiSchema }),
    z.object({ href: hrefSchema }),
    z.lazy(() => z.object({ languages: languagesSchema })),
    z.lazy(() => z.object({ versions: versionsSchema })),
    z.lazy(() => z.object({ dropdowns: dropdownsSchema })),
    z.lazy(() => z.object({ anchors: anchorsSchema })),
    z.lazy(() => z.object({ groups: groupsSchema })),
    z.lazy(() => z.object({ pages: pagesSchema })),
]));
export const nonRecursiveTabSchema = baseTabSchema.and(z.object({ href: hrefSchema }));
export const tabsSchema = z.array(tabSchema).describe('Organizing by tabs');
