// @mintlify/validation/dist/mint-config/schemas/v2/properties/navigation/anchors.js
import { z } from 'zod';
import { colorSchemaWithOptionalLightAndDark } from '../reusable/color.js';
import { hiddenSchema } from '../reusable/hidden.js';
import { hrefSchema } from '../reusable/href.js';
import { iconSchema } from '../reusable/icon.js';
import { openApiSchema } from '../reusable/openapi.js';
import { dropdownsSchema } from './dropdown.js';
import { groupsSchema } from './groups.js';
import { languagesSchema } from './languages.js';
import { pagesSchema } from './pages.js';
import { tabsSchema } from './tabs.js';
import { versionsSchema } from './version.js';
const baseAnchorSchema = z.object({
    anchor: z.string().nonempty().describe('The name of the anchor'),
    icon: iconSchema.optional(),
    color: colorSchemaWithOptionalLightAndDark.optional(),
    hidden: hiddenSchema.optional(),
});
export const anchorSchema = baseAnchorSchema.and(z.union([
    z.object({ openapi: openApiSchema }),
    z.object({ href: hrefSchema }),
    z.lazy(() => z.object({ languages: languagesSchema })),
    z.lazy(() => z.object({ versions: versionsSchema })),
    z.lazy(() => z.object({ dropdowns: dropdownsSchema })),
    z.lazy(() => z.object({ tabs: tabsSchema })),
    z.lazy(() => z.object({ groups: groupsSchema })),
    z.lazy(() => z.object({ pages: pagesSchema })),
]));
export const nonRecursiveAnchorSchema = baseAnchorSchema.and(z.object({ href: hrefSchema }));
export const anchorsSchema = z.array(anchorSchema).describe('Organizing by anchors');
