// @mintlify/validation/dist/mint-config/schemas/v2/properties/navigation/dropdown.js
import { z } from 'zod';
import { hiddenSchema } from '../reusable/hidden.js';
import { hrefSchema } from '../reusable/href.js';
import { iconSchema } from '../reusable/icon.js';
import { openApiSchema } from '../reusable/openapi.js';
import { anchorsSchema } from './anchors.js';
import { groupsSchema } from './groups.js';
import { languagesSchema } from './languages.js';
import { pagesSchema } from './pages.js';
import { tabsSchema } from './tabs.js';
import { versionsSchema } from './version.js';
const baseDropdownSchema = z.object({
    dropdown: z.string().nonempty().describe('The name of the dropdown'),
    icon: iconSchema.optional(),
    hidden: hiddenSchema.optional(),
});
export const dropdownSchema = baseDropdownSchema.and(z.union([
    z.object({ openapi: openApiSchema }),
    z.object({ href: hrefSchema }),
    z.lazy(() => z.object({ languages: languagesSchema })),
    z.lazy(() => z.object({ versions: versionsSchema })),
    z.lazy(() => z.object({ tabs: tabsSchema })),
    z.lazy(() => z.object({ anchors: anchorsSchema })),
    z.lazy(() => z.object({ groups: groupsSchema })),
    z.lazy(() => z.object({ pages: pagesSchema })),
]));
export const nonRecursiveDropdownSchema = baseDropdownSchema.and(z.object({ href: hrefSchema }));
export const dropdownsSchema = z.array(dropdownSchema).describe('Organizing by dropdowns');
